### LIBRARIES #####################################################################################
library(randomForest);
library(argparse);

### ARGUMENTS ####################################################################################
parser <- ArgumentParser();

# if input generated by pipeline, likely named with format *_all_variables.tsv
parser$add_argument('-i', '--input.file', type = 'character', help = 'input file');
parser$add_argument('-o', '--output.directory', type = 'character', help = "output directory");
parser$add_argument('-c', '--cutoff', type = 'double', help = 'cutoff used for rf model of true class', default = 0);
parser$add_argument('-m', '--model.object', type = 'character', help = 'the model');
parser$add_argument('-s', '--sample.name', type = 'character', help = 'sample name');

arguments <- parser$parse_args();

RF <- readRDS(arguments$model.object);

#germlines variants
sample.data=read.table(arguments$input.file,h=T)
#we replace the missing MPOS values by median
vs=as.integer(sample.data[sample.data$MPOS != ".",]$MPOS)
m_mpos_s=median(vs)
sample.data$MPOS=replace(sample.data$MPOS,sample.data$MPOS==".",m_mpos_s)
print(paste0(m_mpos_s))


#now we convert the values
sample.data <- transform(
  sample.data,
  FILE=as.factor(FILE), #sample name
  CHROM=as.factor(CHROM), #Chromosome
  POS=as.integer(POS), # Position
  SNVS=as.factor(SNVS), # join(REFATL) AT TC GA (Signatures)
  BCSQ=as.factor(BCSQ), #Variant impact 
  CODING=as.factor(CODING),#variant is coding
  COSMIC_CENSUS_GENE=as.factor(COSMIC_CENSUS_GENE), #var is in cosmic gene
  COSMIC=as.factor(COSMIC), #var is annotated in cosmic
  GNOMAD=as.integer(GNOMAD), # var is annotated in GNOMAD
  CENTROMER=as.factor(CENTROMER), # var is located in centromeric regions
  MPOS=as.integer(MPOS), # median distance from end of read
  DP=as.integer(DP), # Approximate read depth
  GERMQ=as.integer(GERMQ), # Phred-scaled quality that alt alleles are not germline variants
  SEQQ=as.integer(SEQQ), # Phred-scaled quality that alt alleles are not sequencing errors
  STRANDQ=as.integer(STRANDQ), # Phred-scaled quality of strand bias artifact
  TLOD=as.integer(TLOD), # Log 10 likelihood ratio score of variant existing versus not existing
  AF=AF,
  ADR=as.integer(ADR),
  ADA=as.integer(ADA),
  OR1=as.integer(OR1),
  OR2=as.integer(OR2),
  OA1=as.integer(OA1),
  OA2=as.integer(OA2),
  NS=as.integer(NS), # number of samples with the var
  SOMATIC=as.factor(SOMATIC) #var is somatic or not
)



#sample.features <- sample.data[, colnames(sample.data) %in% names(RF$forest$xlevels)];
sample.features<-subset(sample.data, select = names(RF$forest$xlevels))
names(RF$forest$xlevels)
colnames(sample.data)
colnames(sample.features)
any.missing.features <- which(!names(RF$forest$xlevels) %in% colnames(sample.features));

if (length(any.missing.features) > 0) {
	cat('there are missing features in the input data...\n')
}
#summary(sample.features)
#predicted.class <- predict(RF, sample.features, cutoff = c(1-arguments$cutoff, arguments$cutoff));
predicted.class <- predict(RF, sample.features);

predictions <- data.frame(
	sample.data,
	Predicted.Class = predicted.class
	);

write.table(
	x = predictions,
	file = file.path(arguments$output.directory, paste0(arguments$sample.name, '_predictions.txt')),
	sep = '\t',
	row.names = TRUE,
	quote = FALSE
	);
